### PR TITLE
Add Biome users API routes

### PR DESCRIPTION
**Pull request by dnewh**
Adds the biome and biome-credentials API routes to splinterd behind the corresponding experimental features. This also, enables connecting to a database behind the database experimental feature.

This was tested with the changes to migrations introduced in PR #393

### CODE REVIEW

File: splinterd/Cargo.toml [hunk 1]
```diff
@@ -45,15 +45,19 @@ default = ["generate-certs"]
 stable = ["default"]
 
 experimental = [
-    "circuit-read",
+    "biome",
```
Comment (vaporos) [inline splinterd/Cargo.toml]: Looks like "circuit-read" was removed here, perhaps accidentally.

File: splinterd/src/daemon.rs [hunk 2]
```diff
@@ -382,6 +388,24 @@ impl SplinterDaemon {
...
+        {
+            info!("Adding biome routes");
+            let connection_pool: ConnectionPool =
+                database::create_connection_pool(&self.db_url).unwrap();
+            let mut biome_rest_provider_builder: BiomeRestResourceManagerBuilder =
+                Default::default();
+            biome_rest_provider_builder =
+                biome_rest_provider_builder.with_user_store(connection_pool.clone());
+            #[cfg(feature = "biome-credentials")]
+            {
```
Comment (vaporos) [inline splinterd/src/daemon.rs]: This shouldn't need the block (it is a single statement).

File: splinterd/src/daemon.rs [hunk 3]
```diff
@@ -382,6 +388,24 @@ impl SplinterDaemon {
             rest_api_builder = rest_api_builder.add_resources(circuit_resource.resources());
         }
 
+        #[cfg(feature = "biome")]
+        {
```
Comment (vaporos) [inline splinterd/src/daemon.rs]: When we ever remove this feature, we would have to change the indent of the implementation (or leave a block indented without purpose). Maybe consider making this a function as a result, and calling the function. The call to the function and the function itself can both be guarded by the feature.

File: splinterd/src/daemon.rs [hunk 4]
```diff
@@ -646,6 +671,11 @@ impl SplinterDaemonBuilder {
         self
     }
 
+    pub fn with_db_url(mut self, value: String) -> Self {
```
Comment (vaporos) [inline splinterd/src/daemon.rs]: This should be guarded by the "database" feature (as well as the struct's field).

File: splinterd/src/main.rs [hunk 5]
```diff
@@ -148,6 +148,8 @@ fn main() {
           "If set tls should accept all peer certificates")
         (@arg bind: --("bind") +takes_value
           "Connection endpoint for REST API")
+        (@arg database: --("db") +takes_value
```
Comment (vaporos) [inline splinterd/src/main.rs]: Here we use '--db' but in the configuration file it is 'database'. Suggest we use '--database'/database or '--db-url'/db_url so we have consistency.

File: splinterd/src/main.rs [hunk 6]
```diff
@@ -293,6 +295,13 @@ fn main() {
         .or_else(|| Some("127.0.0.1:8080".to_string()))
         .expect("Must provide a url for REST API endpoint");
 
+    let db_url = matches
```
Comment (vaporos) [inline splinterd/src/main.rs]: This and the arg stuff should all be behind the 'database' feature flag.

File: docker-compose.yaml [hunk 7]
```diff
@@ -31,12 +31,14 @@ services:
...
       args:
         - REPO_VERSION=${REPO_VERSION}
+        - CARGO_ARGS=${CARGO_ARGS}
     entrypoint: |
       bash -c "
         splinterd \
             --generate-certs \
             -c ./configs/splinterd-node-0-docker.toml \
             --registry-file ./node_registries/nodes.yaml \
+            --db postgres://splinter_admin:splinter_test@splinter-db:5432/splinter \
```
Comment (vaporos) [inline docker-compose.yaml]: This becomes problematic if behind a feature flag. We might need to add a SPLINTERD_ARGS env variable so we can pass this through.

File: splinterd/src/daemon.rs (hunk 2 above)
Comment (dnewh) [reply splinterd/src/daemon.rs]: This is a limitation with Rust. Specifically, it throws an error saying attributes cannot be used on expressions.

File: splinterd/Cargo.toml (hunk 1 above)
Comment (dnewh) [reply splinterd/Cargo.toml]: Accident, updated.

File: splinterd/src/daemon.rs (hunk 4 above)
Comment (dnewh) [reply splinterd/src/daemon.rs]: updated

File: splinterd/src/main.rs (hunk 6 above)
Comment (dnewh) [reply splinterd/src/main.rs]: updated

File: docker-compose.yaml (hunk 7 above)
Comment (dnewh) [reply docker-compose.yaml]: added SPLINTERD_ARGS

File: splinterd/src/main.rs (hunk 5 above)
Comment (dnewh) [reply splinterd/src/main.rs]: updated to --database

File: splinterd/src/daemon.rs (hunk 3 above)
Comment (dnewh) [reply splinterd/src/daemon.rs]: Update with the add_biome_routes function. It's a little gross with the memory swapping but it works

File: splinterd/Cargo.toml [hunk 8]
```diff
@@ -45,15 +45,20 @@ default = []
...
     "config-toml",
     "health",
 ]
 
-circuit-read = ["splinter/circuit-read"]
+biome = ["splinter/biome", "database"]
+biome-credentials = ["splinter/biome-credentials", "biome"]
 config-builder = []
 config-toml = ["config-builder"]
+circuit-read = ["splinter/circuit-read"]
```
Comment (peterschwarz) [inline splinterd/Cargo.toml]: Looks like this was moved - it should be before `config-*`

Comment (dnewh) [reply splinterd/Cargo.toml]: rebase issue. Fixed

File: cli/Cargo.toml [hunk 9]
```diff
@@ -26,7 +26,7 @@ experimental = [
 ]
 
 health = ["reqwest", "serde_json"]
-
+biome-credentials = ["database-migrate-biome-credentials"]
```
Comment (vaporos) [inline cli/Cargo.toml]: This should be placed in a separate list below these features, with a comment above it such as "# Aliases - these aliases exist to map libsplinter feature names to their CLI equivalents" (or similar).

File: docker-compose.yaml [hunk 10]
```diff
@@ -31,13 +31,16 @@ services:
       dockerfile: splinterd/Dockerfile-installed-${DISTRO}
       args:
         - REPO_VERSION=${REPO_VERSION}
+        - CARGO_ARGS=${CARGO_ARGS}
```
Comment (vaporos) [inline docker-compose.yaml]: The commit message "Add example splinter db url to splinter-node-0 image." doesn't seem to exactly reflect what is in this commit.

Comment (dnewh) [reply docker-compose.yaml]: updated

File: cli/Cargo.toml (hunk 9 above)
Comment (dnewh) [reply cli/Cargo.toml]: updated

File: splinterd/src/main.rs [hunk 11]
```diff
@@ -147,6 +147,8 @@ fn main() {
           "If set tls should accept all peer certificates")
         (@arg bind: --("bind") +takes_value
           "Connection endpoint for REST API")
+        (@arg database: --("database") +takes_value
```
Comment (peterschwarz) [inline splinterd/src/main.rs]: This will be included if the `"database"` feature is off. It should match the pattern below, like the `generate-certs` block.

File: splinterd/src/main.rs [hunk 12]
```diff
@@ -302,11 +318,20 @@ fn main() {
         .map(String::from)
         .or_else(|| config.registry_file());
 
+    // Allow unused mut for experimental features
+    #[allow(unused_mut)]
+    let mut feature_fields = "".to_string();
```
Comment (peterschwarz) [inline splinterd/src/main.rs]: This is debug output, so it should only be done within the `debug!` macro's call.

You should replace `feature_fields` in the debug log statement with something like:
```rust
        ...
        heartbeat_interval,
        if cfg!(feature = "database") {
            format!("{}, db_url: {}", feature_fields, db_url)
        } else {
            String::new()
        }
    );
```

Comment (dnewh) [reply splinterd/src/main.rs]: updated

File: splinterd/src/daemon.rs [hunk 13]
```diff
@@ -593,6 +605,28 @@ fn start_health_service(
     })?
 }
 
+#[cfg(feature = "biome")]
+fn build_biome_routes(db_url: &str) -> BiomeRestResourceManager {
+    info!("Adding biome routes");
+    let connection_pool: ConnectionPool = database::create_connection_pool(db_url)
+        .map_err(|err| database::DatabaseError::ConnectionError(Box::new(err)))
+        .unwrap();
```
[the remark this answers is not in the document]
Comment (dnewh) [reply splinterd/src/daemon.rs]: handled

File: splinterd/src/main.rs [hunk 14]
```diff
@@ -292,6 +300,14 @@ fn main() {
         .or_else(|| Some("127.0.0.1:8080".to_string()))
         .expect("Must provide a url for REST API endpoint");
 
+    #[cfg(feature = "database")]
+    let db_url = matches
+        .value_of("database")
+        .map(String::from)
+        .or_else(|| config.database())
+        .or_else(|| Some("127.0.0.1:5432".to_string()))
+        .expect("Must provide a url for database");
```
Comment (vaporos) [inline splinterd/src/main.rs]: CLIs should not panic with user-caused errors.

File: splinterd/src/main.rs [hunk 15]
```diff
@@ -164,6 +164,14 @@ fn main() {
             .takes_value(true),
     );
 
+    #[cfg(feature = "database")]
+    let app = app.arg(
+        Arg::with_name("database")
+            .long("database")
+            .long_help("DB connection url, postgres://db_admin:db_password@127.0.0.1:5432/db")
```
Comment (vaporos) [inline splinterd/src/main.rs]: Suggest removing the example URL here; need to keep these short to cut down on wrapping.

Comment (dnewh) [reply splinterd/src/main.rs]: updated to remove example url

File: splinterd/src/main.rs (hunk 14 above)
Comment (vaporos) [reply splinterd/src/main.rs]: This will be handled in a separate PR.

File: cli/Cargo.toml [hunk 16]
```diff
@@ -45,6 +45,9 @@ postgres = [
     "splinter/postgres",
 ]
 
+# Aliases - used to map libsplinter feature names to their CLI counterparts
+biome-credentials = ["database-migrate-biome-credentials"]
```
Comment (peterschwarz) [inline cli/Cargo.toml]: This should be in a separate commit, since it relates only to features of the CLI

Comment (dnewh) [reply cli/Cargo.toml]: Updated
